samples/common/gnss_nmea.c

```diff
@@ -0,0 +1,213 @@
+ * sequence, which will be stripped. For each line, if the line starts with the
+ * letters "AT", ignoring differences in case, then the line will be sent to the
+ * XBee as an AT command. Upon execution of the program, the program will
+ * issue a request to the device to begin streaming NMEA. The user may start
+ * and stop the NMEA stream by typing 'start' or 'stop'.
+ * You may also type 'quit' to exit this program.
+ */
+#include <stdio.h>
+#include <string.h>
+
+#include "xbee/atcmd.h"
+#include "xbee/device.h"
+#include "xbee/gnss.h"
+
+// Common code
+#include "_atinter.h"
+#include "parse_serial_args.h"
+
+// An instance required for the local XBee
+xbee_dev_t my_xbee;
+
+
+int gnss_start_stop_handler( xbee_dev_t *xbee, const void FAR *raw,
+    uint16_t length, void FAR *context);
+
+#define XBEE_FRAME_GNSS_START_STOP_HANDLE     \
+    { XBEE_FRAME_GNSS_START_STOP_RESP, 0, gnss_start_stop_handler, NULL }
+
+int nmea_handler( xbee_dev_t *xbee, const void FAR *raw,
+    uint16_t length, void FAR *context);
+
+#define XBEE_FRAME_GNSS_NMEA_HANDLE     \
+    { XBEE_FRAME_GNSS_RAW_NMEA_RESP, 0, nmea_handler, NULL }
+
+
+const xbee_dispatch_table_entry_t xbee_frame_handlers[] =
+{
+    XBEE_FRAME_HANDLE_LOCAL_AT,
+    XBEE_FRAME_GNSS_START_STOP_HANDLE,
+    XBEE_FRAME_GNSS_NMEA_HANDLE,
+    XBEE_FRAME_TABLE_END
+};
+
+
+// function that handles received GNSS start stop messages
+int gnss_start_stop_handler(xbee_dev_t *xbee, const void FAR *raw,
+    uint16_t length, void FAR *context)
+{
+    const xbee_frame_gnss_start_stop_resp_t FAR *gnss = raw;
+
+    switch (gnss->type) {
+    case XBEE_GNSS_START_RAW_NMEA:
+
+        if (gnss->status != GNSS_START_STOP_STATUS_SUCCESS) {
+            printf("Got failure response for starting GNSS NMEA\n");
+        } else {
+            printf("Got success response for starting GNSS NMEA\n");
+        }
+
+
+        break;
+
+    case XBEE_GNSS_STOP_RAW_NMEA:
+
+        if (gnss->status != GNSS_START_STOP_STATUS_SUCCESS) {
+            printf("Got failure response for stopping GNSS NMEA\n");
+        } else {
+            printf("Got success response for stopping GNSS NMEA\n");
+        }
+
+        break;
+
+    default:
+        printf("Unexpected gnss start stop response type %u\n", gnss->type);
+    }
+
+    return 0;
+}
+
+
+// function that handles received NMEA messages
+int nmea_handler(xbee_dev_t *xbee, const void FAR *raw,
+    uint16_t length, void FAR *context)
+{
+	char nmea_str[1600];
+    const xbee_frame_gnss_raw_nmea_resp_t FAR *gnss = raw;
+
+    int16_t payload_len = length - offsetof(xbee_frame_gnss_raw_nmea_resp_t, payload);
+
+    if (payload_len >= sizeof(nmea_str)) {
+    	printf("PAYLOAD TOO BIG: %lu bytes dropped\n", (payload_len - sizeof(nmea_str) + 1));
+    	payload_len = sizeof(nmea_str) - 1;
+    }
+
+    memcpy(nmea_str, &gnss->payload[0], payload_len);
+    nmea_str[payload_len] = '\0';
+
+    printf("%s\n", nmea_str);
```

Comment:
I think you could get rid of the copy and null termination by using `printf("%.*s", payload_len, nmea_str);`.

Reply:
We still need the `\n` in the printf, but yes.